plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,334 @@
+#include <obs-avc.h>
+#include <obs-module.h>
+#include <util/darray.h>
+#include <util/dstr.h>
+#include <util/pipe.h>
+#include <util/threading.h>
+
+#include "obs-ffmpeg-hls-mux.h"
+#include "obs-ffmpeg-mux.h"
+
+#define do_log(level, format, ...)                       \
+	blog(level, "[ffmpegh hls muxer: '%s'] " format, \
+	     obs_output_get_name(stream->output), ##__VA_ARGS__)
+
+#define warn(format, ...) do_log(LOG_WARNING, format, ##__VA_ARGS__)
+#define info(format, ...) do_log(LOG_INFO, format, ##__VA_ARGS__)
+
+const char *ffmpeg_hls_mux_getname(void *type)
+{
+	UNUSED_PARAMETER(type);
+	return obs_module_text("FFmpegHlsMuxer");
+}
+
+int hls_stream_dropped_frames(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	return stream->dropped_frames;
+}
+
+bool hls_deactivate(struct ffmpeg_muxer *stream, int code)
+{
+	deactivate(stream, code);
+
+	if (stream->mux_thread_joinable) {
+		os_event_signal(stream->stop_event);
+		os_sem_post(stream->write_sem);
+		pthread_join(stream->mux_thread, NULL);
+		stream->mux_thread_joinable = false;
+	}
+
+	pthread_mutex_lock(&stream->write_mutex);
+	for (size_t i = 0; i < stream->mux_packets.num; i++) {
+		struct encoder_packet *packet;
+		packet = &stream->mux_packets.array[i];
+		obs_encoder_packet_release(packet);
+	}
+	da_free(stream->mux_packets);
+	pthread_mutex_unlock(&stream->write_mutex);
+
+	return true;
+}
+
+void ffmpeg_hls_mux_destroy(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+
+	if (stream) {
+		hls_deactivate(stream, 0);
+
+		pthread_mutex_destroy(&stream->write_mutex);
+		os_sem_destroy(stream->write_sem);
+		os_event_destroy(stream->stop_event);
+
+		os_process_pipe_destroy(stream->pipe);
+		dstr_free(&stream->path);
+		dstr_free(&stream->printable_path);
+		dstr_free(&stream->stream_key);
+		dstr_free(&stream->muxer_settings);
+		bfree(data);
+	}
+}
+
+void *ffmpeg_hls_mux_create(obs_data_t *settings, obs_output_t *output)
+{
+	struct ffmpeg_muxer *stream = bzalloc(sizeof(*stream));
+	pthread_mutex_init_value(&stream->write_mutex);
+	stream->output = output;
+
+	/* init mutex, semaphore and event */
+	if (pthread_mutex_init(&stream->write_mutex, NULL) != 0)
+		goto fail;
+	if (os_event_init(&stream->stop_event, OS_EVENT_TYPE_AUTO) != 0)
+		goto fail;
+	if (os_sem_init(&stream->write_sem, 0) != 0)
+		goto fail;
+
+	UNUSED_PARAMETER(settings);
+	return stream;
+
+fail:
+	ffmpeg_hls_mux_destroy(stream);
+	return NULL;
+}
+
+static bool process_packet(struct ffmpeg_muxer *stream)
+{
+	struct encoder_packet *packet;
+	int remaining;
+	bool ret = true;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		ret = write_packet(stream, packet);
```

Comment:
Increment total_bytes by packet->size.

Reply:
Pretty sure this happens in write_packet? Should I do it in process_packet?